tests/test_api.rs

```diff
@@ -72,38 +68,33 @@ fn handle_scope_numbers() {
   }
 }
 
-// TODO: the type checker is kumbaya with this but in reality the
-// `Local<Integer>` created at the end of the test is created in HandleScope
-// `hs2` and not in `hs1` as specified. When this local is accessed, which is
-// after `hs2` is destroyed, a crash happens.
 #[test]
-#[ignore]
-fn handle_scope_early_drop() {
+fn handle_scope_non_lexical_lifetime() {
   let _setup_guard = setup();
   let mut isolate = v8::Isolate::new(Default::default());
-  let mut hs1 = v8::HandleScope::new(&mut isolate);
-  let hs1 = hs1.enter();
+  let scope1 = &mut v8::HandleScope::new(&mut isolate);
+
+  // Despite `local` living slightly longer than `scope2`, this test should
+  // not crash.
   let local = {
-    let mut hs2 = v8::HandleScope::new(hs1);
-    let _hs2 = hs2.enter();
-    v8::Integer::new(hs1, 123)
+    let scope2 = &mut v8::HandleScope::new(scope1);
+    v8::Integer::new(scope2, 123)
   };
   assert_eq!(local.value(), 123);
```

Comment:
The test here has been changed in a rather fundamental way. I agree that the previous version ought to pass (where the local is created with scope1), but I don't agree that the new version should - the local is outliving scope2.

Reply:
Unfortunately Rust lifetimes cannot express the exact semantics of V8 HandlesScopes, so there's always a trade-off to make. The previous version of this test (which used to be disabled because it'd make v8 crash) is now rejected by the borrow checker: https://github.com/denoland/rusty_v8/blob/f6bd4d4a273913443378256683cf94704b66c02b/tests/compile_fail/handle_scope_lifetime_3.rs.